diff --git a/bin/zkEnv.sh b/bin/zkEnv.sh
index 8d93a50c..f1d67b85 100755
--- a/bin/zkEnv.sh
+++ b/bin/zkEnv.sh
@@ -121,6 +121,8 @@ do
    CLASSPATH="$d:$CLASSPATH"
 done
 
+CLASSPATH="$ZOOBINDIR/../../Semantic-Daikon-Checker/target/*:$CLASSPATH"
+
 #make it work for developers
 CLASSPATH="$ZOOBINDIR/../build/classes:$CLASSPATH"
 
diff --git a/bin/zkServer.sh b/bin/zkServer.sh
index 4780118a..3a4461f0 100755
--- a/bin/zkServer.sh
+++ b/bin/zkServer.sh
@@ -158,7 +158,7 @@ start)
     nohup "$JAVA" $ZOO_DATADIR_AUTOCREATE "-Dzookeeper.log.dir=${ZOO_LOG_DIR}" \
     "-Dzookeeper.log.file=${ZOO_LOG_FILE}" "-Dzookeeper.root.logger=${ZOO_LOG4J_PROP}" \
     -XX:+HeapDumpOnOutOfMemoryError -XX:OnOutOfMemoryError='kill -9 %p' \
-    -cp "$CLASSPATH" $JVMFLAGS $ZOOMAIN "$ZOOCFG" > "$_ZOO_DAEMON_OUT" 2>&1 < /dev/null &
+    -cp "$CLASSPATH" $JVMFLAGS -Ddaikon.checkratio=0.99 $ZOOMAIN "$ZOOCFG" > "$_ZOO_DAEMON_OUT" 2>&1 < /dev/null &
     if [ $? -eq 0 ]
     then
       case "$OSTYPE" in
diff --git a/zookeeper-server/pom.xml b/zookeeper-server/pom.xml
index d69e7c63..bd4f7358 100755
--- a/zookeeper-server/pom.xml
+++ b/zookeeper-server/pom.xml
@@ -33,6 +33,13 @@
   <description>ZooKeeper server</description>
 
   <dependencies>
+    <dependency>
+      <groupId>daikon</groupId>
+      <artifactId>Semantic-Daikon-Checker</artifactId>
+      <version>1.0</version>
+      <scope>system</scope>
+      <systemPath>${project.basedir}/../../Semantic-Daikon-Checker/target/Semantic-Daikon-checker-1.0-SNAPSHOT-jar-with-dependencies.jar</systemPath>
+    </dependency>
     <dependency>
       <groupId>com.github.spotbugs</groupId>
       <artifactId>spotbugs-annotations</artifactId>
# diff --git a/zookeeper-server/src/main/java/org/apache/zookeeper/server/DataNode.java b/zookeeper-server/src/main/java/org/apache/zookeeper/server/DataNode.java
# index 8ac8f618..19be832c 100644
# --- a/zookeeper-server/src/main/java/org/apache/zookeeper/server/DataNode.java
# +++ b/zookeeper-server/src/main/java/org/apache/zookeeper/server/DataNode.java
# @@ -39,6 +39,9 @@
#  @SuppressFBWarnings({"EI_EXPOSE_REP", "EI_EXPOSE_REP2"})
#  public class DataNode implements Record {
 
# +    // DIMAS: for testing use
# +    public static DataNode dataNodeInstance = null;
# +
#      // the digest value of this node, calculated from path, data and stat
#      private volatile long digest;
 
# @@ -89,6 +92,8 @@ public DataNode(byte[] data, Long acl, StatPersisted stat) {
#          this.data = data;
#          this.acl = acl;
#          this.stat = stat;
# +
# +        dataNodeInstance = this;
#      }
 
     /**
# diff --git a/zookeeper-server/src/main/java/org/apache/zookeeper/server/DataTree.java b/zookeeper-server/src/main/java/org/apache/zookeeper/server/DataTree.java
# index d3529cf9..0488b62a 100644
# --- a/zookeeper-server/src/main/java/org/apache/zookeeper/server/DataTree.java
# +++ b/zookeeper-server/src/main/java/org/apache/zookeeper/server/DataTree.java
# @@ -94,6 +94,8 @@
#   */
#  public class DataTree {
 
# +    public static DataTree dataTreeInstance = null;
# +
#      private static final Logger LOG = LoggerFactory.getLogger(DataTree.class);
 
#      private final RateLogger RATE_LOGGER = new RateLogger(LOG, 15 * 60 * 1000);
# @@ -307,6 +309,8 @@ public DataTree() {
#              LOG.error("Unexpected exception when creating WatchManager, exiting abnormally", e);
#              ServiceUtils.requestSystemExit(ExitCode.UNEXPECTED_ERROR.getValue());
#          }
# +
# +        dataTreeInstance = this;
#      }
 
     /**
diff --git a/zookeeper-server/src/main/java/org/apache/zookeeper/server/PrepRequestProcessor.java b/zookeeper-server/src/main/java/org/apache/zookeeper/server/PrepRequestProcessor.java
old mode 100644
new mode 100755
index 399f9488..536b0c8d
--- a/zookeeper-server/src/main/java/org/apache/zookeeper/server/PrepRequestProcessor.java
+++ b/zookeeper-server/src/main/java/org/apache/zookeeper/server/PrepRequestProcessor.java
@@ -18,6 +18,9 @@
 
 package org.apache.zookeeper.server;
 
+import daikon.DaikonLogger;
+import daikon.GlobalState;
+import daikon.InvariantChecker;
 import java.io.ByteArrayOutputStream;
 import java.io.IOException;
 import java.io.StringReader;
@@ -765,6 +768,20 @@ protected void pRequest(Request request) throws RequestProcessorException {
         // request.type + " id = 0x" + Long.toHexString(request.sessionId));
         request.setHdr(null);
         request.setTxn(null);
+        LOG.info("DIMAS: " + GlobalState.mode.toString());
+        switch (GlobalState.mode)
+        {
+            case TEST:
+                //CHANG: we expose states at the entry of all requests
+                DaikonLogger.getLogger().dumpStates();
+                break;
+            case PRODUCTION:
+                InvariantChecker.assertAll();
+                break;
+            case ASSERTING:
+                LOG.trace("Asserting, so skip another triggered");
+                break;
+        }
 
         try {
             switch (request.type) {
diff --git a/zookeeper-server/src/main/java/org/apache/zookeeper/server/ZooKeeperServer.java b/zookeeper-server/src/main/java/org/apache/zookeeper/server/ZooKeeperServer.java
old mode 100644
new mode 100755
index d1365e48..6852d9ae
--- a/zookeeper-server/src/main/java/org/apache/zookeeper/server/ZooKeeperServer.java
+++ b/zookeeper-server/src/main/java/org/apache/zookeeper/server/ZooKeeperServer.java
@@ -18,6 +18,11 @@
 
 package org.apache.zookeeper.server;
 
+import daikon.DaikonLogger;
+import daikon.GlobalState;
+import daikon.Pair;
+import java.lang.reflect.Field;
+import java.util.*;
 import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
 import java.io.ByteArrayOutputStream;
 import java.io.File;
@@ -93,6 +98,9 @@
  */
 public class ZooKeeperServer implements SessionExpirer, ServerStats.Provider {
 
+    //CHANG: for testing use
+    public static ZooKeeperServer zksInstance = null;
+
     protected static final Logger LOG;
 
     public static final String GLOBAL_OUTSTANDING_LIMIT = "zookeeper.globalOutstandingLimit";
@@ -350,6 +358,8 @@ public ZooKeeperServer(FileTxnSnapLog txnLogFactory, int tickTime, int minSessio
             getClientPortListenBacklog(),
             txnLogFactory.getDataDir(),
             txnLogFactory.getSnapDir());
+
+        zksInstance = this;
     }
 
     public String getInitialConfig() {
@@ -683,6 +693,8 @@ public synchronized void startup() {
 
         setState(State.RUNNING);
 
+        // GlobalState.mode = GlobalState.T2CMode.PRODUCTION;
+
         requestPathMetricsCollector.start();
 
         localSessionEnabled = sessionTracker.isLocalSessionsEnabled();
